### PR TITLE
Implement GUI Autofocus checkbox 

### PR DESCRIPTION
**Pull request by K4rishma**
**Checked on Simulation**
Create autofocus chekbox iwhich runs autofocus in the given roi and provides fixed number of focus points to tiled acquisition. 

This can be only merged after the below PR has been merged.
https://github.com/delmic/odemis/pull/2342

### CODE REVIEW

File: src/odemis/acq/stitching/_tiledacq.py [hunk 1]
```diff
@@ -904,3 +904,180 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+
+    def __init__(self, streams, stage, area, focus, ccd, future, overlap=0.2, settings_obs=None, log_path=None,
+                 zlevels=None,
+                 registrar=REGISTER_GLOBAL_SHIFT, weaver=WEAVER_MEAN, focusing_method=FocusingMethod.NONE):
+        # site and feature means the same
+        self._stage = stage
+        self._future = future
+        self.streams = streams
+        self.area = area
+        self._rois = [area, area]  # list of roi # TODO ask GUI to provide a list of rois
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:952]: @Eric this task is linked to the future which is called by on_acquire() in win/acquisition.py This tasks accepts list of rois for which autofocus and tiled acquisition will run for each roi.

My question (also added as a TODO in the code) is that if creating list of roi is a part of this PR?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 2]
```diff
@@ -904,3 +904,180 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+        self._ccd = ccd  # TODO to delete after ccd.data components is used from streams in do_autofocus_roi
+        self._focus = focus
+        self.focus_rng = self._focus.getMetadata().get(model.MD_POS_ACTIVE_RANGE, None)
+
+        if self.focus_rng is None:
+            # METEOR is running
+            self.focus_rng = self._focus.getMetadata().get(model.MD_SAFE_REL_RANGE, None)
+            if self.focus_rng is not None:
+                self.focus_rng = (self.focus_rng[0] + self._focus.position.value,
+                                  self.focus_rng[1] + self._focus.position.value)
```
Comment (pieleric) [inline src/odemis/acq/stitching/_tiledacq.py:962]: `self._focus.position.value` -> `self._focus.position.value["x"]`
(and same for y)

File: src/odemis/acq/stitching/_tiledacq.py (hunk 1 above)
Comment (pieleric) [reply src/odemis/acq/stitching/_tiledacq.py:952]: Just take as input "areas", and use it here.

File: src/odemis/acq/stitching/_tiledacq.py [hunk 3]
```diff
@@ -904,3 +904,180 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
     return future
+
+
+def acquireOverview(streams, stage, area, focus, ccd, overlap=0.2, settings_obs=None, log_path=None, zlevels=None,
+                    registrar=REGISTER_GLOBAL_SHIFT, weaver=WEAVER_MEAN, focusing_method=FocusingMethod.NONE):
+    """
+    Start a tiled acquisition task for the given streams (SEM or FM) in order to
+    build a complete view of the TEM grid. Needed tiles are first acquired for
+    each stream, then the complete view is created by stitching the tiles.
+
```
Comment (pieleric) [inline src/odemis/acq/stitching/_tiledacq.py:915]: docstring for all the input arguments is needed.

File: src/odemis/acq/align/roi_autofocus.py [hunk 4]
```diff
@@ -119,22 +126,29 @@ def autofocus_in_roi(
         ccd: model.Component,
         focus: model.Component,
         focus_range: tuple,
-        nx: int = 3,
-        ny: int = 3,
+        n_focus_points: tuple = (3, 3),
```
Comment (K4rishma) [inline src/odemis/acq/align/roi_autofocus.py:129]: Modify the test cases for this function once the earlier pull requests with test cases gets merged

File: src/odemis/acq/stitching/_tiledacq.py [hunk 5]
```diff
@@ -890,6 +891,7 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
     future.running_subf = model.InstantaneousFuture()
     future._task_lock = threading.Lock()
     # Create a tiled acquisition task
+    # TODO input the focus_range as input argument and do not calculate it in this task
     task = TiledAcquisitionTask(streams, stage, area, overlap, settings_obs, log_path, future=future, zlevels=zlevels,
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:895]: add focus range in tiled acquisition and adjust the code withing the tiled acquisition class to use the focus range from the input arguments

File: src/odemis/acq/align/roi_autofocus.py [hunk 6]
```diff
@@ -119,22 +126,29 @@ def autofocus_in_roi(
         ccd: model.Component,
```
Comment (K4rishma) [inline src/odemis/acq/align/roi_autofocus.py:126]: Comment from another PR, Thera "autofocus_roi.py takes the current position of the focus component as the starting position, it might be useful to clarify this in the autofocus_roi docstring". -

File: src/odemis/acq/stitching/_tiledacq.py [hunk 7]
```diff
@@ -670,8 +672,6 @@ def _get_triangulated_focus_point(self, x, y):
                 break
 
         if point_in_triangle is False:
-            # TODO determine a better way for dealing with points outside of the focused area, for instance
-            #  by using linear extrapolation.
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:674]: Can you put the TODO back?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 8]
```diff
@@ -781,7 +780,7 @@ def _adjustFocus(self, das, i, ix, iy):
                                                   good_focus=self._good_focus,
                                                   rng_focus=self._focus_rng,
                                                   method=MTD_EXHAUSTIVE)
-            _, focus_pos, _ = self._future.running_subf.result()  # blocks until autofocus is finished
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:784]: the `, _` is from the confidence, you should not remove it

**Review by tepals — CHANGES_REQUESTED** [review]
just did a very quick review, it looks like you introduced some errors during rebasing

File: src/odemis/acq/stitching/_tiledacq.py [hunk 9]
```diff
@@ -926,3 +928,194 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+            future._task_state = CANCELLED
+            future.running_subf.cancel()
+            logging.debug("acquisition overview cancelled.")
+        return True
+
+    def estimate_time(self, roi_idx=0, actual_time_per_roi=None):
+        """
+        Estimates the time for the rest of the acquisition.
+
+        :param site_idx: (int) index of the current site
```
Comment (pieleric) [inline src/odemis/acq/stitching/_tiledacq.py:1026]: The name of the param is different.
Also I think this is actually the number of RoI's already *acquired*. That means when the first RoI is acquired, you pass 1, (not 0 as the description implies).

File: src/odemis/acq/stitching/_tiledacq.py [hunk 10]
```diff
@@ -926,3 +928,194 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+
+                # run autofocus for the selected roi
+                self._future.running_subf = autofocus_in_roi(roi,
+                                                             self._stage,
+                                                             self._ccd,
+                                                             self._focus,
+                                                             self.focus_rng,
+                                                             self.n_focus_points,
+                                                             self.conf_level)
+
```
Comment (pieleric) [inline src/odemis/acq/stitching/_tiledacq.py:1065]: Each check for cancel + start of a sub-future should be within a lock block (`with self._task_lock`).
This is the whole point of the lock: to avoid the possibility that someone cancels just before the start of the sub-future. With the lock, we always ensure that the task state is either checked before starting the sub-future, or the sub future is started, and it's immediately cancelled.

File: src/odemis/acq/stitching/_tiledacq.py [hunk 11]
```diff
@@ -926,3 +928,194 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+                    raise
+
+                logging.debug(f"acquisition overview is completed for roi number {idx} with {roi} values")
+
+                # cancel the sub future
+                if self._future._task_state == CANCELLED:
+                    raise CancelledError()
+
+                # Store the actual time during acquisition one roi
+                actual_time_per_roi = time.time() - start_time
```
Comment (pieleric) [inline src/odemis/acq/stitching/_tiledacq.py:1107]: I don't think there will be a lot of variations, but instead of using just the duration of the latest RoI acquired, you could do the average: move `start_time` to before the for loop, and then compute as:
```suggestion
                actual_time_per_roi = (time.time() - start_time) / (idx + 1)
```

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/acq/align/roi_autofocus.py [hunk 12]
```diff
@@ -52,38 +52,43 @@ def do_autofocus_in_roi(
...
                 # run autofocus
                 f._running_subf = align.AutoFocus(ccd, None, focus, rng_focus=focus_range)
-                foc_pos, foc_lev, conf_measure = f._running_subf.result(timeout=900)
-                if foc_lev >= conf_level:
+                foc_pos, foc_lev, conf = f._running_subf.result(timeout=900)
+                if conf >= conf_level:
                     focus_positions.append([stage.position.value["x"],
                                             stage.position.value["y"],
                                             focus.position.value["z"]])
                     logging.debug(f"Added focus with confidence of {foc_lev} at position: {focus_positions[-1]}")
```
Comment (pieleric) [inline src/odemis/acq/align/roi_autofocus.py:79]:
```suggestion
                    logging.debug(f"Added focus with confidence of {conf} at position: {focus_positions[-1]}")
```

File: src/odemis/acq/stitching/_tiledacq.py [hunk 13]
```diff
@@ -926,3 +928,194 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+            # No absolute range is available, so use a relative range (e.g. as on METEOR
+            self.focus_rng = self._focus.getMetadata().get(model.MD_SAFE_REL_RANGE, SAFE_REL_RANGE_DEFAULT)
+            self.focus_rng = (self.focus_rng[0] + self._focus.position.value["z"],
+                              self.focus_rng[1] + self._focus.position.value["z"])
+
+        self.conf_level = 0.8
+        self.focusing_method = focusing_method
+        # The number of focus points are constant for now
+        # TODO: make it variable according to the size of the ROI, is it useful?
+        self.n_focus_points = (3, 3)
```
[the remark this answers is not in the document]
Comment (K4rishma) [reply src/odemis/acq/stitching/_tiledacq.py:1000]: This task of scaling the number of focus points with the fov is in the backlog. I have not worked on it. Is it something to prioritize?

File: src/odemis/acq/stitching/_tiledacq.py (hunk 9 above)
Comment (K4rishma) [reply src/odemis/acq/stitching/_tiledacq.py:1026]: I passed 0 for the first site as the index number will be zero for it.

File: src/odemis/gui/win/acquisition.py [hunk 14]
```diff
@@ -1061,14 +1066,31 @@ def on_acquire(self, evt):
...
 
-        self.acq_future = stitching.acquireTiledArea(acq_streams, self._main_data_model.stage, area=self.area,
-                                                     overlap=self.overlap,
-                                                     settings_obs=self._main_data_model.settings_obs,
-                                                     log_path=self.filename_tiles,
-                                                     weaver=WEAVER_MEAN,
-                                                     registrar=REGISTER_IDENTITY,
-                                                     zlevels=zlevels,
-                                                     focusing_method=focus_mtd)
+        if self.autofocus_roi_ckbox.value:
```
[the remark this answers is not in the document]
Comment (K4rishma) [reply src/odemis/gui/win/acquisition.py:1069]: Should I do it in this PR, or will it be there in your PR for overview acquisition in mimas?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 15]
```diff
@@ -141,8 +145,7 @@ def __init__(self, streams, stage, area, overlap, settings_obs=None, log_path=No
 
             # used in re-focusing method
             self._focus_points = numpy.array(focus_points) if focus_points else None
-            # triangulate focus points
-            self._tri_focus_points = Delaunay(self._focus_points[:, :2]) if focus_points else None
+            self._tri_focus_points = Delaunay(self._focus_points[:, :2]) if focus_points else None # triangulate focus points
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:148]: I think this is more readable with the comment above the line, why did you move it?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 16]
```diff
@@ -649,10 +654,10 @@ def _acquireTiles(self):
 
     def _get_z_on_focus_plane(self, x, y):
         if not self._focus_plane:
-            gamma, normal = linalg.fit_plane_lstsq(self._focus_points)
-            self._focus_plane["gamma"] = gamma
+            c, normal = linalg.fit_plane_lstsq(self._focus_points)
+            self._focus_plane["c"] = c
             self._focus_plane["normal"] = normal
-        point_on_plane = (0, 0, self._focus_plane["gamma"])  # where the plane intersects with the z-axis
+        point_on_plane = (0, 0, self._focus_plane["c"])  # where the plane intersects with the z-axis
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:660]: To keep it consistent with `linalg.fit_plane_lstsq`, can you name it back to `gamma`?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 17]
```diff
@@ -926,3 +928,198 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+
+def acquireOverview(streams, stage, area, focus, ccd, overlap=0.2, settings_obs=None, log_path=None, zlevels=None,
+                    registrar=REGISTER_GLOBAL_SHIFT, weaver=WEAVER_MEAN, focusing_method=FocusingMethod.NONE):
+    """
+    Start autofocus and tiled acquisition tasks for the given list of areas.
+
+    :param streams: (list of Stream) the streams to acquire
+    :param stage: (Actuator) the stage to move
+    :param area: (list of area) area is defined by 4 floats :left, top, right, bottom positions
+     of the acquisition area (in m)
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:941]: for readability it's nice to have an indent for multi-line docstring argument explanations

Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:941]: this explanation is not very clear, is this the list of areas you want to acquire? If yes, can you update the description to explain that? And does it then make sense to rename the argument from `area` to `areas`?

File: src/odemis/acq/stitching/_tiledacq.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tepals) [reply src/odemis/acq/stitching/_tiledacq.py:1000]: this is something we still need to test, could be that it's every 4 tiles, but it could also be that 3x3 tiles is always good. I think it's fine to keep it in the backlog for now, until the first testing has been done.

File: src/odemis/acq/stitching/_tiledacq.py [hunk 18]
```diff
@@ -926,3 +928,198 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+                # run autofocus for the selected roi
+                self._future.running_subf = autofocus_in_roi(roi,
+                                                             self._stage,
+                                                             self._ccd,
+                                                             self._focus,
+                                                             self.focus_rng,
+                                                             self.n_focus_points,
+                                                             self.conf_level)
+
+                logging.debug(f"Autofocus is running for roi number {idx} with {roi} values")
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:1067]: `roi` is a bounding box of the overview-area containing the left, top, right, bottom positions of this area. I think this debug message can be a bit more clear:
```suggestion
                logging.debug(f"Autofocus is running for roi number {idx}, with bounding box: {roi} [m]")
```

File: src/odemis/acq/stitching/_tiledacq.py [hunk 19]
```diff
@@ -926,3 +928,198 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+
+                try:
+                    # TODO use the "estimated time * 3 + 1" as timeout error
+                    focus_points = self._future.running_subf.result()
+                except Exception:
+                    logging.debug(
+                        f"Autofocus within roi failed for roi number {idx} with {roi} values")
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:1074]: same:
`roi` is a bounding box of the overview-area containing the left, top, right, bottom positions of this area. I think this debug message can be a bit more clear:
```suggestion
                logging.debug(f"Autofocus is running for roi number {idx}, with bounding box: {roi} [m]")
```

File: src/odemis/acq/stitching/_tiledacq.py [hunk 20]
```diff
@@ -926,3 +928,198 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+                                                             self._ccd,
+                                                             self._focus,
+                                                             self.focus_rng,
+                                                             self.n_focus_points,
+                                                             self.conf_level)
+
+                logging.debug(f"Autofocus is running for roi number {idx} with {roi} values")
+
+                try:
+                    # TODO use the "estimated time * 3 + 1" as timeout error
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:1070]: What do you mean with this TODO? Is this not something you can add very quickly now?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 21]
```diff
@@ -926,3 +928,198 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+            future.running_subf.cancel()
+            logging.debug("acquisition overview cancelled.")
+        return True
+
+    def estimate_time(self, roi_idx=0, actual_time_per_roi=None) -> float:
+        """
+        Estimates the time for the rest of the acquisition.
+
+        :param roi_idx: (int) index of the current roi
+        :param actual_time_per_roi: (float) the actual time spent for the current roi
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:1027]: this is not the actual time spent for the current roi, but the average time spent for all rois acquired until now.

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/acq/stitching/_tiledacq.py (hunk 20 above)
Comment (K4rishma) [reply src/odemis/acq/stitching/_tiledacq.py:1070]: There was an earlier comment which said "maybe use a  timeout error" in this block. That comment is updated to this TODO. I am not sure if using timeout error will be very useful, I can maybe remove the TODO?

Comment (tepals) [reply src/odemis/acq/stitching/_tiledacq.py:1073]: I think you could just directly implement it, something like the following (I did not test this):
```python
t = actual_time_per_roi * 3 + 1
focus_points = self._future.running_subf.result(t)
except  TimeoutError:
    logging.warning("Failed to run autofocus for roi number {idx}, with bounding box: {roi} [m]")
    self._future.running_subf.cancel()
except ....
```

File: src/odemis/acq/stitching/_tiledacq.py [hunk 22]
```diff
@@ -926,3 +930,198 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
...
+                                                             self.n_focus_points,
+                                                             self.conf_level)
+
+                logging.debug(f"Autofocus is running for roi number {idx}, with bounding box: {roi} [m]")
+
+                try:
+                    # TODO use the "estimated time * 3 + 1" as timeout error
+                    focus_points = self._future.running_subf.result()
+                except Exception:
+                    logging.debug(f"Autofocus is running for roi number {idx}, with bounding box: {roi} [m]")
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:1076]: This log message does not make sense here, it is exactly the same as the one on line 1070. Don't you want to log something about the exception here?

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/gui/win/acquisition.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/win/acquisition.py:1068]: I'll push the changes with my PR.

File: src/odemis/acq/stitching/_tiledacq.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/acq/stitching/_tiledacq.py:1004]: Agree. Let's make it a separate task. I think that if the ROI is tiny, 3x3 is probably too many.

**Review by pieleric — APPROVED** [review]
